src/test/java/jep/test/TestSharedInterpreter.java

```diff
@@ -86,6 +92,53 @@ public static void testSharedTypes() throws Throwable{
         }
     }
 
+    public static void testSharedJPyObject() throws Throwable {
+        try (Interpreter interp = new SharedInterpreter()) {
+            PyObject list = interp.getValue("[]", PyObject.class);
+            Thread t = new Thread(() -> {
+                try (Interpreter interp2 = new SharedInterpreter()) {
+                    list.getAttr("append", PyCallable.class).call("test");
+                    interp2.set("l", list);
+                    interp2.exec("l.append('test2')");
+                }
+            });
+            t.start();
+            t.join();
+            String test2 = list.getAttr("pop", PyCallable.class).callAs(String.class);
+            String test = list.getAttr("pop", PyCallable.class).callAs(String.class);
+            if (!test.equals("test")) {
+                throw new IllegalStateException("Expecting 'test', not " + test); 
+            }
+            if (!test2.equals("test2")) {
+                throw new IllegalStateException("Expecting 'test2', not " + test2); 
+            }
+            /* Make sure it doesn't work on threads without interpreters or in SubInterpreters. */
+            boolean[] pass = { true };
+            t = new Thread(() -> {
+                try {
+                    list.getAttr("append", PyCallable.class).call("test");
+                    pass[0] = false;
+                } catch (JepException e) {
+                }
+                try (Interpreter interp2 = new SubInterpreter()) {
+                    list.getAttr("append", PyCallable.class).call("test");
+                    pass[0] = false;
+                } catch (JepException e) {
```

Comment:
What is the message on the exception that is thrown?  It's not clear to me how you made it work for SharedInterpreters but not SubInterpreters.

Reply:
The message is "Invalid thread access" from line 122 of MemoryManager.

The MemoryManager is keeping track of which interpreters are sharing memory and uses a ThreadLocal to track which threads those interpreters are valid on. When PyPointer or PyObject gets the thread state for any native operation it is now going through the MemoryManager which will return the thread state if there is an interpreter for that MemoryManager on that thread. If there is no interpreter for a memory manager then the "Invalid thread access" is thrown.

The implementation of the MemoryManager and PyPointer do not specifically behave differently for SubInterpreter vs. SharedInterpreter, But during construction the SharedInterpreters wil use a static MemroyManager that is shared and SubInterpreters will construct a new MemoryManager and this difference in construction causes different behavior.